types/scene/wlr_scene.c

```diff
@@ -220,6 +232,119 @@ void wlr_scene_rect_set_color(struct wlr_scene_rect *rect, const float color[sta
 	scene_node_damage_whole(&rect->node);
 }
 
+struct wlr_scene_buffer *wlr_scene_buffer_create(struct wlr_scene_node *parent,
+		struct wlr_buffer *buffer) {
+	struct wlr_scene_buffer *scene_buffer = calloc(1, sizeof(*scene_buffer));
+	if (scene_buffer == NULL) {
+		return NULL;
+	}
+	scene_node_init(&scene_buffer->node, WLR_SCENE_NODE_BUFFER, parent);
+
+	scene_buffer->buffer = wlr_buffer_lock(buffer);
+
+	scene_node_damage_whole(&scene_buffer->node);
+
+	return scene_buffer;
+}
+
+void wlr_scene_buffer_set_source_box(struct wlr_scene_buffer *scene_buffer,
+		const struct wlr_fbox *box) {
+	struct wlr_fbox *cur = &scene_buffer->src_box;
+	if ((wlr_fbox_empty(box) && wlr_fbox_empty(cur)) ||
+			(box != NULL && memcmp(cur, box, sizeof(*box)) == 0)) {
+		return;
+	}
+
+	if (box != NULL) {
+		memcpy(cur, box, sizeof(*box));
+	} else {
+		memset(cur, 0, sizeof(*cur));
+	}
+
+	scene_node_damage_whole(&scene_buffer->node);
+}
+
+void wlr_scene_buffer_set_dest_size(struct wlr_scene_buffer *scene_buffer,
+		int width, int height) {
+	if (scene_buffer->dst_width == width && scene_buffer->dst_height == height) {
+		return;
+	}
+
+	scene_node_damage_whole(&scene_buffer->node);
+	scene_buffer->dst_width = width;
+	scene_buffer->dst_height = height;
+	scene_node_damage_whole(&scene_buffer->node);
+}
+
+void wlr_scene_buffer_set_transform(struct wlr_scene_buffer *scene_buffer,
+		enum wl_output_transform transform) {
+	if (scene_buffer->transform == transform) {
+		return;
+	}
+
+	scene_node_damage_whole(&scene_buffer->node);
+	scene_buffer->transform = transform;
+	scene_node_damage_whole(&scene_buffer->node);
+}
+
+static struct wlr_texture *scene_buffer_get_texture(
+		struct wlr_scene_buffer *scene_buffer, struct wlr_renderer *renderer) {
+	struct wlr_client_buffer *client_buffer =
+		wlr_client_buffer_get(scene_buffer->buffer);
+	if (client_buffer != NULL) {
+		return client_buffer->texture;
+	}
+
+	if (scene_buffer->texture != NULL) {
+		return scene_buffer->texture;
+	}
+
+	scene_buffer->texture =
+		wlr_texture_from_buffer(renderer, scene_buffer->buffer);
+	if (scene_buffer->texture == NULL) {
```

Comment:
Is this check really necessary?

Reply:
Yes -- the name might not be well-chosen but `wlr_texture_from_buffer` will import/upload the buffer. This can fail.

Reply:
Ah, sorry, that's not what I meant. Returning NULL if the value is NULL is redundant if said value is returned otherwise.

Reply:
Oh, right, fixed!